src/core/lib/security/transport/security_handshaker.c

```diff
@@ -223,17 +228,23 @@ static grpc_error *on_handshake_next_done_locked(
         GRPC_ERROR_CREATE_FROM_STATIC_STRING("Handshake failed"), result);
   }
   // Send data to peer.
-  grpc_slice to_send = grpc_slice_from_copied_buffer(
-      (const char *)bytes_to_send, bytes_to_send_size);
-  grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
-  grpc_slice_buffer_add(&h->outgoing, to_send);
-  grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
-                      &h->on_handshake_data_sent_to_peer);
-  // If handshake has completed, check peer and so on.
+  if (bytes_to_send_size > 0) {
+    grpc_slice to_send = grpc_slice_from_copied_buffer(
+        (const char *)bytes_to_send, bytes_to_send_size);
+    grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
+    grpc_slice_buffer_add(&h->outgoing, to_send);
+    grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
+                        &h->on_handshake_data_sent_to_peer);
+  }
+  // If handshake has completed, check peer and so on. Otherwise, need to read
+  // more data from the peer.
   if (handshaker_result != NULL) {
     GPR_ASSERT(h->handshaker_result == NULL);
     h->handshaker_result = handshaker_result;
     error = check_peer_locked(exec_ctx, h);
+  } else {
+    grpc_endpoint_read(exec_ctx, h->args->endpoint, h->args->read_buffer,
```

Comment:
I was not aware of the edge case where the handshake is complete but we still have bytes to send to the peer.  Given that, I think that even without the read code-path, we still have the same problem with multiple pending callbacks, because both the check-peer and write callbacks will both be pending at the same time.

I would strongly prefer to have only a single pending callback at any one time, because it greatly simplifies the error-handling logic.  If you can't find a way to do that, then we will need to discuss taking an additional ref to the handshaker and changing the error-handling logic to handle that additional ref somehow, and I would really prefer to avoid that complexity.

@ctiller has confirmed for me that calling `grpc_endpoint_write()` with a 0-length buffer is very inefficient and not really intended to work.  So we should continue to avoid that if possible.